### DISCUSSION TITLE
How to generate a x.509 certificate with a Serial Number using openssl?

### DISCUSSION THREAD

**Discussion by Vishwamithra37** [MatthiasValvekens/pyHanko#106] [Q&A] [answered]
So, I tried almost all the available methods on the internet, set_Serial number, create_serial number and other flags. But I'm unable to get a certificate with a serial number. And make it visible in the certificate field during signing of PDF using pyHanko.
Any help will be greatly appreaciated. :)
`openssl req -newkey ec:<(openssl ecparam -name secp384r1) \
            -x509 \
            -sha256 \
            -days 3650 \
            -nodes \
            -out example3.crt \
            -keyout example3.key \
            -set_serial 8787 \
            -subj "/CA=65216895787/C=IN/ST=Telangana`
I'm currently try thins, but it is not showing the serial number field. 
On the same note, how to access specific fields in a cert using pyHanko, like if I want to access the ST field in my certificate, or the public key, is it possible? How?

- > Question: How to extract the certificate fields seperately from a signature? That is, get the certificate.

**Comment by MatthiasValvekens** [comment, marked as answer]
You can supply your own interpolation values for text stamps / qr stamps using the [`appearance_text_params` argument to `sign_pdf` and friends](https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.sign.signers.pdf_signer.html#pyhanko.sign.signers.pdf_signer.PdfSigner.sign_pdf). You could use those to create `%(st)s` and `%(serial)s` parameters, if you'd want to.

As for how to extract the information to populate those variables from the certificate: have a look at the properties on [this class](https://github.com/wbond/asn1crypto/blob/8a4c621e34f8cbdc83b047aa617e66dcbfde0f75/asn1crypto/x509.py#L2144) in `asn1crypto` (that's also the one pyHanko uses internally). Getting the serial number is straightforward, but extracting a RDN segment like `ST` will require a bit more digging. You'll want to parse that out of the `x509.Name` structure that represents the certificate's subject. `asn1crypto` can help you with that as well.

**Reply by Vishwamithra37** [reply]
No like, I'm 100% sure pyHanko has this functionality, it displays the certificate details during the valistion code. So, umm, there is no other way but to use that library to parse the certificate indvidual data?

**Reply by MatthiasValvekens** [reply]
`asn1crypto` is a dependency of pyHanko, and the certificate objects that pyHanko passes around in its internal APIs are instances of `x509.Certificate`. In other words, you don’t have to install any additional libs :)
